src/composer.ts

```diff
@@ -0,0 +1,491 @@
+  decodeError?: Error;
+}
+
+export enum AtomicTransactionComposerStatus {
+  /** The atomic group is still under construction. */
+  BUILDING,
+
+  /** The atomic group has been finalized, but not yet signed. */
+  BUILT,
+
+  /** The atomic group has been finalized and signed, but not yet submitted to the network. */
+  SIGNED,
+
+  /** The atomic group has been finalized, signed, and submitted to the network. */
+  SUBMITTED,
+
+  /** The atomic group has been finalized, signed, submitted, and successfully committed to a block. */
+  COMMITTED,
+}
+
+/** A class used to construct and execute atomic transaction groups */
+export class AtomicTransactionComposer {
+  /** The maximum size of an atomic transaction group. */
+  static MAX_GROUP_SIZE: number = 16;
+
+  private status = AtomicTransactionComposerStatus.BUILDING;
+  private transactions: TransactionWithSigner[] = [];
+  private methodCalls: Map<number, ABIMethod> = new Map();
+  private signedTxns: Uint8Array[] = [];
+  private txIDs: string[] = [];
+
+  /**
+   * Get the status of this composer's transaction group.
+   */
+  getStatus(): AtomicTransactionComposerStatus {
+    return this.status;
+  }
+
+  /**
+   * Get the number of transactions currently in this atomic group.
+   */
+  count(): number {
+    return this.transactions.length;
+  }
+
+  /**
+   * Create a new composer with the same underlying transactions. The new composer's status will be
+   * BUILDING, so additional transactions may be added to it.
+   */
+  clone(): AtomicTransactionComposer {
+    const theClone = new AtomicTransactionComposer();
+
+    theClone.transactions = this.transactions.map(({ txn, signer }) => ({
+      // not quite a deep copy, but good enough for our purposes (modifying txn.group in buildGroup)
+      txn: Transaction.from_obj_for_encoding({
+        ...txn.get_obj_for_encoding(),
+        // erase the group ID
+        grp: undefined,
+      }),
+      signer,
+    }));
+    theClone.methodCalls = new Map(this.methodCalls);
+    theClone.signedTxns = this.signedTxns.slice();
+    theClone.txIDs = this.txIDs.slice();
+
+    return theClone;
+  }
+
+  /**
+   * Add a transaction to this atomic group.
+   *
+   * An error will be thrown if the transaction has a nonzero group ID, the composer's status is
+   * not BUILDING, or if adding this transaction causes the current group to exceed MAX_GROUP_SIZE.
+   */
+  addTransaction(txnAndSigner: TransactionWithSigner): void {
+    if (this.status !== AtomicTransactionComposerStatus.BUILDING) {
+      throw new Error(
+        'Cannot add transactions when composer status is not BUILDING'
+      );
+    }
+
+    if (this.transactions.length === AtomicTransactionComposer.MAX_GROUP_SIZE) {
+      throw new Error(
+        `Adding an additional transaction exceeds the maximum atomic group size of ${AtomicTransactionComposer.MAX_GROUP_SIZE}`
+      );
+    }
+
+    if (txnAndSigner.txn.group && txnAndSigner.txn.group.some((v) => v !== 0)) {
```

Comment:
Yes, since `groupID = hash(hash(txn1) + hash(txn2) + ...))`, if any of the transactions already has a group ID, its hash will be taken with that group ID. So after we replace that with the generated group ID, the whole group won't verify properly anymore.